### PR TITLE
Zip data descriptor validation update and AppContext switch option

### PR DESCRIPTION
**Pull request by buyaa-n**

### CODE REVIEW

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs [hunk 1]
```diff
@@ -189,5 +190,15 @@ private static bool SeekBackwardsAndRead(Stream stream, byte[] buffer, out int b
                 return true;
             }
         }
+
+        internal static bool GetSuppressValidation()
+        {
+            if (AppContext.TryGetSwitch("System.Compression.SuppressValidation", out bool suppressValidation))
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipHelper.cs]: Naming.  This should be `System.IO.Compression.ZipArchive.SuppressValidation` or  `System.IO.Compression.ZipArchiveEntry.SuppressValidation`.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs [hunk 2]
```diff
@@ -128,6 +128,30 @@ public static async Task ZipArchiveEntry_InvalidUpdate(string zipname)
...
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Fix not shipped for .NET Framework.")]
+        public static async Task ZipArchiveEntry_BypassValidation()
+        {
+            MemoryStream stream = await LocalMemoryStream.readAppFileAsync(zfile("normal.zip"));
+            PatchDataRelativeToFileName(Encoding.ASCII.GetBytes(s_tamperedFileName), stream, 8);  // patch uncompressed size in file header
+
+            using (ZipArchive archive = new ZipArchive(stream, ZipArchiveMode.Read))
+            {
+                // Disable the flag variable using reflection as corresponding AppContext value is cached
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: This should be tested using RemoteExecutor so that the static state doesn't leak into other processes.  Also, it should ideally be done using the SetSwitch API so as not to depend on internals.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 3]
```diff
@@ -482,44 +482,62 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
...
                     return false;
                 }
 
                 reader.BaseStream.Seek(extraFieldLength + entry.CompressedLength, SeekOrigin.Current); // seek to end of compressed file from which Data descriptor starts
-                uint dataDescriptorSignature = reader.ReadUInt32();
-                bool wasDataDescriptorSignatureRead = false;
-                int seekSize = 0;
-                if (dataDescriptorSignature == DataDescriptorSignature)
+                uint[] buffer = new uint[6];
+                // let's try to match the smallest possible structure (3 x 4) 32 bit without signature
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Does it make sense to skip these if the size from central dir is greater than Unit32.MaxValue?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 4]
```diff
@@ -482,44 +482,62 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
...
                 {
                     return false;
                 }
 
                 reader.BaseStream.Seek(extraFieldLength + entry.CompressedLength, SeekOrigin.Current); // seek to end of compressed file from which Data descriptor starts
-                uint dataDescriptorSignature = reader.ReadUInt32();
-                bool wasDataDescriptorSignatureRead = false;
-                int seekSize = 0;
-                if (dataDescriptorSignature == DataDescriptorSignature)
+                uint[] buffer = new uint[6];
```
Comment (eerhardt) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: You can just `stackalloc` this. No reason to allocate an array for this small of a value.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 5]
```diff
@@ -534,6 +552,24 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
 
             return true;
         }
+
+        internal static ulong ConvertToUlong(uint loverAddressValue, uint higherAddressValue)
```
Comment (eerhardt) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: this can be `private`

Comment (eerhardt) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: `loverAddressValue` => `lowerAddressValue`

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 6]
```diff
@@ -534,6 +552,24 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
...
+        {
+            return checked((ulong)loverAddressValue + (((ulong)higherAddressValue) << 32));
```
Comment (eerhardt) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Is `checked` really necessary here? I don't think anything here can overflow.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 7]
```diff
@@ -482,44 +482,62 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
...
-
-                if (is64bit)
+                //let's try to match the 64 bit structures 64 bit without signature
+                buffer[4] = reader.ReadUInt32();
+                seekSize += 4;
+                if (TestMatch(entry, DataDescriptorSignature, buffer[0], ConvertToUlong(buffer[1], buffer[2]), ConvertToUlong(buffer[3], buffer[4])))
                 {
-                    compressedSize = reader.ReadInt64();
-                    uncompressedSize = reader.ReadInt64();
+                    compressedSize = (long)ConvertToUlong(buffer[1], buffer[2]);
```
Comment (eerhardt) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Can we cache `ConvertToUlong(buffer[1], buffer[2])` and `ConvertToUlong(buffer[3], buffer[4])` in a local variable, so we don't have to do it twice?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 3 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: By `skip these` you mean skip validation?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 7 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Sure i was thinking to changes it

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 6 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: right

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs [hunk 8]
```diff
@@ -482,44 +482,62 @@ public static bool TryValidateBlock(BinaryReader reader, ZipArchiveEntry entry)
...
+                if (TestMatch(entry, buffer[0], buffer[1], buffer[2], buffer[3]))
+                {
+                    compressedSize = buffer[2];
+                    uncompressedSize = buffer[3];
+                    goto MatchFound;
+                }
 
-                if (reader.BaseStream.Length < reader.BaseStream.Position + seekSize)
+                // At this point prior to trying to match 64 bit structures we need to make sure that version is high enough
+                if (entry._versionToExtract < ZipVersionNeededValues.Zip64)
```
Comment (buyaa-n) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: If we remove this validation `@MikeCodesDotNET`'s zip would pass

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 5 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: > loverAddressValue => lowerAddressValue

Thas is from WPF 😛

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: I have tried with `AppContext.SetSwitch(...)` worked only if run the individual test on VS, but not work for bulk test run from console, anyways will try with that `RemoteExecutor` that might help

Comment (ericstj) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: You need to make sure the call to SetSwitch happens before the static constructor that reads the setting.  You can take a look for the various places that use AppContext and their tests to identify the best pattern for testing.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 3 above)
Comment (ericstj) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: I mean skip reading as 32-bits if you know the sizes cannot be represented as 32-bits.  You’d still validate but only consider 64-but format.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_InvalidParametersAndStrangeFiles.cs]: AppContext.SetSwitch(...) worked well with RemoteExecutor.Invoke(), thanks!

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs (hunk 3 above)
Comment (buyaa-n) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.cs]: Ah I see, @eerhardt actually come up with a similar solution https://github.com/dotnet/runtime/compare/master...eerhardt:FixZipIssues.
So now the question is if we want to go with @eerhardt solution + AppContext switch or WPF solution + AppContext switch. WPF solution worked with the 17GB zip locally, but didn't added a test for that, too big, the zip entry alone is 3GB of compressed size.

For me to avoid for any other strange edge case we might better use WPF solution which has more forgiving logic than testing exact matching case we have now. If we remove this condition on row 516 even `@MikeCodesDotNET`'s buggy zip will be passed and we might not need to add AppContext switch to disable validation
